pandas/core/internals/blocks.py

```diff
@@ -663,22 +663,25 @@ def replace_list(
         if is_string_dtype(values.dtype):
             # Calculate the mask once, prior to the call of comp
             # in order to avoid repeating the same computations
-            mask = ~isna(values)
-            masks = [
-                compare_or_regex_search(values, s[0], regex=regex, mask=mask)
+            na_mask = ~isna(values)
+            masks = (
+                extract_bool_array(
+                    compare_or_regex_search(values, s[0], regex=regex, mask=na_mask)
+                )
                 for s in pairs
-            ]
+            )
         else:
             # GH#38086 faster if we know we dont need to check for regex
-            masks = [missing.mask_missing(values, s[0]) for s in pairs]
+            masks = (
+                extract_bool_array(missing.mask_missing(values, s[0])) for s in pairs
```

Comment:
is there a chance of an API change here (not necessarily a bad one) with something like

```
src_list = [1, 2, 3]
dest_list = [2, 3, 4]
values = np.array([1, 2, 3])
```

IIUC in the status quo we end up with `[2, 3, 4]` but with this we'd end up with all-4s?

Reply:
Ah, looks like this would be invalid for inplace=True. Good catch!

I'll also add a test case, since it looks like this isn't caught by CI.